src/dstack/_internal/core/services/ssh/tunnel.py

```diff
@@ -1,77 +1,175 @@
+import abc
+import asyncio
 import os
 import shlex
 import subprocess
 import tempfile
-from typing import Dict, Optional
+from dataclasses import dataclass
+from typing import Dict, Iterable, List, Optional, Tuple
 
 from dstack._internal.core.errors import SSHError
 from dstack._internal.core.models.instances import SSHConnectionParams
 from dstack._internal.core.services.ssh import get_ssh_error
 from dstack._internal.utils.logging import get_logger
-from dstack._internal.utils.path import PathLike
+from dstack._internal.utils.path import FilePath, FilePathOrContent, PathLike
 
 logger = get_logger(__name__)
 SSH_TIMEOUT = 15
+SSH_DEFAULT_OPTIONS = {
+    "StrictHostKeyChecking": "no",
+    "UserKnownHostsFile": "/dev/null",
+    "ExitOnForwardFailure": "yes",
+    "StreamLocalBindUnlink": "yes",
+    "ConnectTimeout": "3",
+}
+
+
+class Socket(abc.ABC):
+    @abc.abstractmethod
+    def render(self) -> str:
+        pass
+
+
+@dataclass
+class UnixSocket(Socket):
+    path: PathLike
+
+    def render(self) -> str:
+        return str(self.path)
+
+
+@dataclass
+class TCPSocket(Socket):
```

Comment:
<sup>terminological nitpicking</sup>

I think `IPSocket` would be a more correct name, since `Unix` in `UnixSocket` means _socket_family_ (`AF_UNIX` for Unix domain sockets, `AF_INET` for IPv4 sockets and `AF_INET6` for IPv6 sockets). while `TCP` means _protocol_[^1]. Although SSH without additional software is only capable of tunneling TCP traffic, I wouldn't make a data object representing “just” IP protocol address be aware of this detail.

[^1]: `AF_UNIX` sockets do not have protocols, but they do have `socket_type`, both `SOCK_STREAM` (≈ `IPPROTO_TCP` IP protocol)  and `SOCK_DGRAM` (≈ `IPPROTO_UDP`).

See: [`socket(2)`](https://man7.org/linux/man-pages/man2/socket.2.html), [`unix(7)`](https://man7.org/linux/man-pages/man7/unix.7.html), [`ip(7)`](https://man7.org/linux/man-pages/man7/ip.7.html).